### PR TITLE
scripts: add simple upmerge helper script

### PR DESCRIPTION
**Pull request by karhama**
Minimal script for upmerging Zephyr project.

Signed-off-by: Kari Hamalainen <kari.hamalainen@nordicsemi.no>

### CODE REVIEW

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Putting a -1 on for now. I have some basic questions:

- what portions of our upmerge process does this intend to automate, and what portions are out of scope and will be left manual?
- I am assuming you have given some thought about how not to break the oss-history automatic rebase steps, right?
- who will maintain this?
- what are the proposed processes for making sure this does not drift from our internal documentation?

Feel free to point me at an internal bug tracker reference if you prefer to answer there.

**Comment by karhama** [comment]
> Putting a -1 on for now. I have some basic questions:
> 
> * what portions of our upmerge process does this intend to automate, and what portions are out of scope and will be left manual?
> * I am assuming you have given some thought about how not to break the oss-history automatic rebase steps, right?
> * who will maintain this?
> * what are the proposed processes for making sure this does not drift from our internal documentation?
> 
> Feel free to point me at an internal bug tracker reference if you prefer to answer there.

Target is to have MVP for automatic upmerge process. This script is quite close to ncs-compare but instead of only printing the information this script will do reverting (fromlist / fromtree commits) and then merge upstream. This script will pass only if there are no conflicts (in reverting or in merging).

Idea is to trigger this tool using e.g. github actions and if it passes execute some CI tests and even automatically upmerge. This way we will at least receive frequent feedback of issues vs. current manual upmerge process.

This is WIP mainly because I'm not sure how to test other than Zephyr project (I have only run "west ncs-upmerger zephyr"). I'm not even sure if there are any other project which it can be tested at this point.

**Comment by mbolivar-nordic** [comment]
> This is WIP mainly because I'm not sure how to test other than Zephyr project (I have only run "west ncs-upmerger zephyr"). I'm not even sure if there are any other project which it can be tested at this point.

Give it a shot with mcuboot first before moving on to other projects, I would guess would be easiest.

**Comment by carlescufi** [comment]
@mbolivar-nordic can you please re-review?
@de-nordic can you take a look?

**Review by mbolivar-nordic — APPROVED** [review]
Seems harmless; let's try it

**Review by de-nordic — APPROVED** [review]
Looks OK to me.

**Review by thst-nordic — APPROVED** [review]

File: scripts/west_commands/ncs_commands.py [hunk 1]
```diff
@@ -481,6 +481,77 @@ def allowed_project(self, zp):
             log.msg(status, color=log.WRN_COLOR)
             likely_merged(np, zp, nsha, zsha)
 
+class NcsUpmerger(NcsWestCommand):
```
[the remark this answers is not in the document]
Comment (karhama) [reply scripts/west_commands/ncs_commands.py:484]: done

**Pull request merged**